src/data-access/services/round-robin-ftp-client-pool.ts

```diff
@@ -0,0 +1,113 @@
+import { ConnectionStatus } from '../enums/connection-status'
+import { FtpClient } from '../interfaces/ftp-client'
+import { FileMetadata } from '../value-objects/file-metadata'
+
+const MAX_FAILED_OPERATIONS_THRESHOLD: number = 20
+const SUCCESSFUL_OPERATION_WEIGHT: number = 1
+const FAILED_OPERATION_WEIGHT: number = 2
+
+export class RoundRobinFtpClientPool implements FtpClient {
+  private connectedFtpClient?: FtpClient
+  private onConnectionStatusChangedCallback?: (connectionStatus: ConnectionStatus) => void
+  private failedOperationsTracker: number = 0
+
+  public constructor(private ftpClients: readonly FtpClient[]) {}
+
+  public async connect(): Promise<void> {
+    await this.getConnectedFtpClient()
+  }
+
+  private async getConnectedFtpClient(): Promise<FtpClient> {
+    if (!this.connectedFtpClient || !this.connectedFtpClient.isConnected()) {
+      this.connectedFtpClient = await this.getFtpClientByRoundRobin()
+    }
+    return this.connectedFtpClient
+  }
+
+  private async getFtpClientByRoundRobin(): Promise<FtpClient> {
+    await this.disconnect()
+    this.emitConnectionStatus(ConnectionStatus.CONNECTING)
+    for (const ftpClient of this.ftpClients) {
+      try {
+        await ftpClient.connect()
+        this.emitConnectionStatus(ConnectionStatus.CONNECTED)
+        ftpClient.setOnConnectionStatusChangedCallback(connectionStatus => this.emitConnectionStatus(connectionStatus))
+        return ftpClient
+      } catch {
+        ftpClient.setOnConnectionStatusChangedCallback(() => {})
+      }
+    }
+    this.emitConnectionStatus(ConnectionStatus.DISCONNECTED)
+    throw new Error('No available iNews servers.')
+  }
+
+  private emitConnectionStatus(connectionStatus: ConnectionStatus): void {
+    this.onConnectionStatusChangedCallback?.(connectionStatus)
+  }
+
+  public isConnected(): boolean {
+    return this.connectedFtpClient?.isConnected() ?? false
+  }
+
+  public async changeWorkingDirectory(path: string): Promise<void> {
+    return this.trackFailedOperations(async () => {
+      const ftpClient: FtpClient = await this.getConnectedFtpClient()
+      await ftpClient.changeWorkingDirectory(path)
+    })
+  }
+
+  private async trackFailedOperations<T>(operationCallback: () => Promise<T>): Promise<T> {
+    await this.reconnectIfHasTooManyFailedOperations()
+    try {
+      const result: T = await operationCallback()
+      this.failedOperationsTracker = Math.max(0, this.failedOperationsTracker - SUCCESSFUL_OPERATION_WEIGHT)
+      return result
```

Comment:
Can i suggest to just get rid of the const assignment here if it is not used and isn't typed in a way that is needed for understanding the logic? 😄 

```suggestion
      this.failedOperationsTracker = Math.max(0, this.failedOperationsTracker - SUCCESSFUL_OPERATION_WEIGHT)
      return await operationCallback()
```

Reply:
It is needed. We await the operationCallback, such that we only update the failed operations tracker if the called operation was successful :)

Reply:
Ok :D Was just hard to see immediately that there was a side-effect of the success or failure of operationCallback :)